### PR TITLE
NVMe QoS Benchmarking Tool

### LINKED ISSUE #45: Comprehensive benchmarks
Basically as the title describes. We need a way to validate that future changes to the scheduler behaviour meaningfully improve performance.

### PR DESCRIPTION
**Pull request by BrandonPacewic**
Closes #45.

### Summary

  Adds a purpose-built benchmarking tool (`tools/nvme-qos-bench/`) to validate and measure the NVMe QoS scheduler's effect on p99 latency, throughput, and CPU overhead under controlled mixed-workload conditions.

  - `tools/nvme-qos-bench/nvme_qos_bench.py` — CLI with six subcommands:
    - check — verify system readiness (root, fio, sysfs, debugfs stats).
    - run -c <A-E> — run a condition profile, interleaving QoS-off and QoS-on iterations across queue depths and weights.
    - analyze — print latency/throughput tables, 95% CIs, Welch's t-test p-values, and kernel dispatch counters from a completed run.
    - validate — quick functional smoke test (~2-3 min) confirming sysfs, namespace policy, and ioprio classification.
    - conditions — list or inspect condition profiles.
    - list — enumerate past runs with commit, branch, condition, and pass/fail summary.
  - Five condition profiles covering the scheduler's core mechanisms:

  | ID  | Name                     | Test Target                                                    |
  |-----|--------------------------|------------------------------------------------------------------|
  | A   | Zero-overhead baseline   | Confirms no overhead at QD1/QD4 with no contention               |
  | B   | Few queues, high density | WRR arbitration under maximum per-queue contention               |
  | C   | Device saturation        | Primary operating target: 2:8 high:normal job ratio, 256K writes |
  | D   | Majority high-prio       | Credit exhaustion; normal traffic still serviced                 |
  | E   | Weight sweep             | Proportional fairness across weights 1, 4, 9, 19, 99             |

  - Statistical analysis — means, stddev, 95% CIs (pure-Python t-distribution), and Welch's t-test p-values implemented without external numeric dependencies.
  - Kernel stats integration — reads per-queue dispatch counters from debugfs (`CONFIG_NVME_QOS_STATS=y`) and includes them in aggregate output and reports.

### Reviewer Checklist

Kernel changes:

  - [ ] CONFIG_NVME_QOS_STATS correctly depends on both NVME_QOS and DEBUG_FS; enabling it without both should not compile.
  - [ ] All NQS_INC() call sites on the hot path (nvme_qos_kick, enqueue, dispatch) are guarded by #ifdef CONFIG_NVME_QOS_STATS and add no overhead when disabled.
  - [ ] The debugfs seq_file show function holds no locks while reading atomic64 counters — verify no lock ordering issue with sq_lock.
  - [ ] nvme_qos_debugfs_root and per-device qos_debugfs_dir are created and destroyed symmetrically with driver init/remove (no leak on error path or module unload).
  - [ ] stats_reset write handler is safe to call concurrently with the scheduler hot path (atomic writes are fine; verify no non-atomic fields are reset).
  - [ ] All sq_lock acquisitions in the new stats/debugfs code use spin_lock_irqsave().

  Benchmarking tool (tools/nvme-qos-bench/)

 - [ ] validate fails (not silently passes) when CONFIG_NVME_QOS_STATS is absent — counter-based assertions must not be vacuously true.
  - [ ] Shared-baseline matching in _store_qos_results correctly reuses one baseline entry across all weights in condition E without
  cross-contaminating results from different depths.
  - [ ] Drift probe threshold (post / pre > 3.0) does not fire spuriously on cold drives or after a write-heavy condition.
  - [ ] SLC/FTL degraded-iteration detection (70% of median) uses the right metric per condition — condition D has no io_write_ios (randread normal
  workload).
  - [ ] device.restore_state() covers all state mutated during a run: QoS enabled, weight, namespace policy, max depth.
  - [ ] Welch's t-test _betai -> _betacf continued fraction converges correctly at n=5 iterations.
  - [ ] Fairness validator handles the edge case where high demand exactly equals the weight share (boundary of demand-limited vs weight-limited regime).
  - [ ] fio timeout (runtime * 2 + 60s) is sufficient for the longest condition runtime.

### CODE REVIEW

**Review by phannawich — CHANGES_REQUESTED** [review]
TL;DR: this is more of a black box, therefore, I will not approve the PR

After trying to figure out on how to handle this pull request, I will have to automatically reject this Pull Request. Even thought your code might work correctly, it is defeat the purpose of reviewing a pull request, so I cannot approve it.

It is impossible for me to understand all the code and be able to check if this pull request is working correctly based on the issue open earlier.

For the future Pull Request, the code that have ridiculous amount of line (more than 2000) will automatically get reject by me. 

What I suggest is that you make the problem smaller. Submit each part at a time with the correlated problem that the PR is trying to solve (Or in this case, that each PR is trying to measure). I will read and double check PR with the issues. If code behavior is solving what address in the issue and I will approve it.

**Comment by phannawich** [comment]
P.S. the amount of line of code is def not a good measurement on that fact that PR is good or bad. But this is still very hard for me to ensure that the code is working correctly.

File: tools/nvme-qos-bench/lib/analysis.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    if df <= 0:
+        return float('inf')
+    if df in T_CRITICAL_95:
+        return T_CRITICAL_95[df]
+    # Find closest
+    keys = sorted(T_CRITICAL_95.keys())
+    for i, k in enumerate(keys):
+        if k > df:
+            if i == 0:
+                return T_CRITICAL_95[keys[0]]
```
Comment (Benjamin-Anderson-II) [inline tools/nvme-qos-bench/lib/analysis.py]: From what I can see this branch condition isn't ever going to be met with the values present in `T_CRITICAL_95`. There is no integer that exists between 0 and 1(the lowest key value present)...

If this is for future proofing in case of a different set of critical values, then I think a comment explaining that would be nice.

Feel free to correct me if I'm wrong though

File: tools/nvme-qos-bench/lib/analysis.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            # Linear interpolation
```
Comment (Benjamin-Anderson-II) [inline tools/nvme-qos-bench/lib/analysis.py]: Worth noting that this is an _approximation_ of the in-between critical values, as they are not linearly distributed.

File: tools/nvme-qos-bench/lib/analysis.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    med_writes = sorted(write_ios)[len(write_ios) // 2]
+    med_iops = sorted(iops_vals)[len(iops_vals) // 2]
+
+    degraded = []
+    for i in range(len(iterations)):
+        if med_writes > 0 and write_ios[i] < 0.7 * med_writes:
+            degraded.append(i)
+        elif med_iops > 0 and iops_vals[i] < 0.7 * med_iops and util_vals[i] > 90:
+            degraded.append(i)
```
Comment (Benjamin-Anderson-II) [inline tools/nvme-qos-bench/lib/analysis.py]: Correct me if I'm wrong, but `write_ios`, `iops_vals`, and `util_vals` all only contain positive values right?

If so, remove the `med_writes > 0` and `med_iops > 0` checks, since the value within will never be strictly less than 0, so the branch condition will never be met. I don't know how much speed up it'll bring, but hopefully some.

```py
for i in range(len(iterations)):
    if write_ios[i] < 0.7 * med_writes:
        degraded.append(i)
    elif iops_vals[i] < 0.7 * med_iops and util_vals[i] > 90
        degraded.append(i)
```
